src/Microsoft.DotNet.Tools.Compiler/Program.cs

```diff
@@ -319,16 +322,49 @@ private static bool CompileProject(ProjectContext context, string configuration,
                     if (projectDependency.Project.Files.SourceFiles.Any())
                     {
                         var projectOutputPath = GetProjectOutput(projectDependency.Project, projectDependency.Framework, configuration, outputPath);
-                        compilerArgs.Add($"--reference:{projectOutputPath}");
+                        references.Add(projectOutputPath);
                     }
                 }
                 else
                 {
-                    compilerArgs.AddRange(dependency.CompilationAssemblies.Select(r => $"--reference:{r.ResolvedPath}"));
+                    references.AddRange(dependency.CompilationAssemblies.Select(r => r.ResolvedPath));
                 }
+
                 compilerArgs.AddRange(dependency.SourceReferences);
             }
 
+            compilerArgs.AddRange(references.Select(r => $"--reference:{r}"));
+
+            var runtimeContext = ProjectContext.Create(context.ProjectDirectory, context.TargetFramework, new[] { RuntimeIdentifier.Current });
+            var libraryExporter = runtimeContext.CreateExporter(configuration);
+
+            if (compilationOptions.PreserveCompilationContext == true)
+            {
+                var dependencyContext = DependencyContextBuilder.FromLibraryExporter(
+                    libraryExporter, context.TargetFramework.DotNetFrameworkName, context.RuntimeIdentifier);
+
+                var writer = new DependencyContextWriter();
+                var depsJsonFile = Path.Combine(intermediateOutputPath, context.ProjectFile.Name + "dotnet-compile.deps.json");
+                using (var fileStream = File.Create(depsJsonFile))
+                {
+                    writer.Write(dependencyContext, fileStream);
+                }
+
+                compilerArgs.Add($"--resource:\"{depsJsonFile}\",DependencyContext.json");
+
+                var refsFolder = Path.Combine(outputPath, "refs");
```

Comment:
Create an issue for removing the `refs` folder from compiled output. We should be able to resolve them from the cache at runtime in this scenario. When publishing we will want to copy to a flat folder.